File: tree/102. Binary Tree Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
//  Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

// For example:
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7
   
//  return its level order traversal as:
// [
//   [3],
//   [9,20],
//   [15,7]
// ]  
// 題意: 要把一個樹中同一個高度的數字都放進一個陣列中
// 再將這些陣列依照高度放在一個陣列中
// 要注意的是，要先加左再加右
// 像範例如果回傳
// [
//   [3],
//   [20,9],
//   [7,15]
// ]  
// 就會測不過

// 思路:
// 用一個物件去存
// level 當 key ，value 是 array
// 依序將同level 放進這個array
// 最後再把key 去掉，依序存進一個array 就好

class TreeNode {
    constructor(val, left, right, level) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
        this.level = (level === undefined ? 0 : level)
    }
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
    if(!root){
        return []
    }
    root.level = 0;
    let levelRecord = {};
    calLevel(root, root.level, levelRecord);
    let ans = Object.keys(levelRecord).map(level => levelRecord[level]);
    return ans;

};

/**
 *
 *
 * @param {TreeNode} node
 * @param {Number} level
 * @param {*} levelRecord
 */
function calLevel(node, level, levelRecord) {
    levelRecord[level] = levelRecord[level] || [];
    levelRecord[level].push(node.val);
    if (node.left) {
        calLevel(node.left, level + 1, levelRecord);
    }
    if (node.right) {
        calLevel(node.right, level + 1, levelRecord);
    }
  
}

File: tree/102. Binary Tree Level Order Traversal.cpp

/*
Given the root of a binary tree, 
return the level order traversal of its nodes' values. 
(i.e., from left to right, level by level).
題意：要你用一個二維容器，按照樹的level，把同level的值存在一個容器裡
返回一個二維容器 

思路：一個輔助函式，帶有當前節點、二維容器和level數
逐層去收集每一層的節點

Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: [[3],[9,20],[15,7]]

Example 2:
Input: root = [1]
Output: [[1]

Example 3:
Input: root = []
Output: []

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> vvi;
        bfs(root,vvi,0);
        return vvi;
    }
    void bfs(TreeNode* node,vector<vector<int>> &vvi,int level){
        if(node == nullptr){
            return;
        }
        if(vvi.size() == level) vvi.push_back(vector<int>());
        vvi[level].push_back(node->val);
        bfs(node->left,vvi,level+1);
        bfs(node->right,vvi,level+1);
    }
};

//Runtime 4 ms Beats 80.61% Memory 13.3 MB Beats 12.83%